File: lib/scripts/chiffres.dart
// ! Ce code a été repris du code de Kenji qui se trouve dans python/chiffres.py

import 'dart:math';

import 'package:math_expressions/math_expressions.dart';

const petits = [
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10,
  1,
  2,
  3,
  4,
  5,
  6,
  7,
  8,
  9,
  10
];

const grands = [
  25,
  50,
  75,
  100
];

// Tirage de 6 chiffres aléatoires et d'une cible
List tirageChiffres() {
  final Random random = Random();
  final List<int> tousLesNombres = [...petits, ...grands];
  tousLesNombres.shuffle(random);
  
  final List<int> tirage = tousLesNombres.sublist(0, 6);
  final int cible = random.nextInt(900) + 100; 
  
  return [cible, tirage];
}

// Utilisation de la librairie math_expression pour calculer un string
int calculer(String expression) {
  Parser p = Parser();
  Expression exp = p.parse(expression);
  ContextModel cm = ContextModel();
  return (exp.evaluate(EvaluationType.REAL, cm)).round();
}

File: python/chiffres.py
# Ici vous retrouverez le script python fait par Kenji pour la partie chiffre.

# Sa version en Dart se trouve dans lib/scripts/chiffres.dart

import random
import re

PETITS = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10] * 2
GRANDS = [25, 50, 75, 100]

def tirage_chiffres():
    tous_les_nombres = PETITS + GRANDS
    random.shuffle(tous_les_nombres)
    tirage = tous_les_nombres[:6]
    cible = random.randint(100, 999)
    return tirage, cible

def utiliser_uniquement_nombres(chaine, nombres):
    chiffres_utilises = re.findall(r'\d+', chaine)
    temp = nombres[:]
    try:
        for c in chiffres_utilises:
            n = int(c)
            if n in temp:
                temp.remove(n)
            else:
                return False
        return True
    except:
        return False

def jeu_chiffres():
    print("=== Jeu des chiffres ===")
    tirage, cible = tirage_chiffres()
    print("Nombres tirés :", tirage)
    print("Nombre à atteindre :", cible)

    solution = input("Entrez votre solution (ex: (100 + 25) * 3) : ")

    if not utiliser_uniquement_nombres(solution, tirage):
        print("Votre solution utilise des nombres non autorisés.")
        return

    try:
        resultat = eval(solution)
        if not isinstance(resultat, (int, float)):
            print("Résultat invalide.")
            return
        difference = abs(cible - resultat)
        print(f"Résultat obtenu : {resultat}")
        if difference == 0:
            print("Bravo ! Cible atteinte.")
        else:
            print(f"Écart par rapport à la cible : {difference}")
    except Exception as e:
        print("Erreur dans l'expression :", e)

jeu_chiffres()